ynput/ayon-backend
# Settings: Migrate settings based on bundle

## PR Checklist

- [x] migrate settings by bundle function
- [x] migrateSettingsByBundle endpoint
- [x] extend save bundle endpoint to allow settings migration in one step 
- [x] handle site settings properly
- [x] dispatch settings changed events
- [ ] use in-method caching of settings to reduce query amount
- [x] log everything

## Testing

### Phase 1: Test migrateSettingsByBundle endpoint

- Turn off developer mode
- Create `OldBundle` bundle with an older version of an addon and change its studio and project settings
- Create `NewBundle` bundle with a newer version of the same addon and keep it untouched
- Set `OldBundle` to production and `NewBundle` to staging

Call (using postman, curl or httpie) a `[POST]` request with administrator or service privileges to `/api/migrateSettingsByBundle` with the following JSON payload

```json
{
  "sourceBundle": "OldBundle",
  "targetBundle": "NewBundle",
  "sourceVariant": "production",
  "targetVariant": "staging"
}
```

- Check the logs for errors. `Failed to validate XXXX with value XXX` warnings are expected if the target addon does not have migration method in place. that setting won't be migrated.
- Ensure the settings were correctly propagated in studio settings and project settings

- When https://github.com/ynput/ayon-backend/pull/311 is merged, use "promote bundle" function on the staging bundle and ensure site settings were also migrated
